### DISCUSSION TITLE
Tool library

### DISCUSSION THREAD

**Discussion by mattjoneswo** [terjeio/ioSender#107] [Ideas]
Is there anything in the works/progress to move the Tools Library from the old Gcode Sender to the new IOSender?    Would be nice to predefine tools that can be called up via M6 T(x), then probing would update the Z length/offset.

**Comment by terjeio** [comment]
> Is there anything in the works/progress to move the Tools Library from the old Gcode Sender to the new IOSender?

There was no tool library in the old sender code, just an interface for handling tool tables in the controller. This has been ported over but is not used/tested much. If your controller is [grblHAL](https://github.com/grblHAL) the tool table can be enabled by editing [grbl/config.h](https://github.com/grblHAL/core/blob/master/config.h), search for `N_TOOLS`. If enabled then the corresponding tool tab will automatically show up in the sender.

A tool table/library stored in the sender is not something I plan to implement.

An afterthought:
What I have is a kind of tool library that I use for my CO2 laser, for mapping tool numbers to laser settings. This is not a part of ioSender, it is for a custom sender that I have not published (yet). Perhaps this is what you saw in the old library code?

**Reply by mattjoneswo** [reply]
Ah.  I remember seeing the N_Tools in the config and trying to follow it unsuccessfully (my limitations).   I will explore this option.  I remember seeing a screenshot of the Gcode Sender before your iosender with a “tools” tab.  Perhaps this will solve my problem. When I’m testing a program with M6 tool change (M6 T1 / M6 T2) I am able to successfully move machine to G59.3 position and probe tool after changing tool manually, yet the new length is not being set/updated and program resumes at original Z-referenced height.   

-side quest: do you have a patreon?
